README
NAME
    ElasticSearch::Transport::AEHTTP - AnyEvent::HTTP backend for
    ElasticSearch

VERSION
    version 0.06

SYNOPSIS
        use ElasticSearch;
        my $e = ElasticSearch->new(
            servers     => 'search.foo.com:9200',
            transport   => 'aehttp',
        );

        # blocking request
        $e->cluster_health->recv;

        # non-blocking request
        $e->cluster_health->cb( sub {
            if ($@) {
                log "An error occurred: $@";
            } else {
                my $response = shift;
                log $response;
            }
        });
        AE::cv->recv();

        # fire-and-forget vs scoped
        {
            $e->delete_index(index => 'foo');
            $e->delete_index(index => 'bar')->cb( sub { print "Done"});
            my $cv = $e->delete_index(index=>'baz');
        }
        AE::cv->recv;
        #   - foo and bar will be deleted
        #   - baz will not be deleted

DESCRIPTION
    ElasticSearch::Transport::AEHTTP uses AnyEvent::HTTP to talk to
    ElasticSearch asynchronously over HTTP.

USING AEHTTP
    Any request to ElasticSearch returns an AnyEvent::CondVar. You have
    three options for how you use them:

  Blocking
        $cv = $e->cluster_health;
        $result = $cv->recv;

    When you call "recv()" on a CondVar, your program will block until that
    CondVar is ready to return a value.

    If an error was thrown, then "recv()" will "die". You will need to wrap
    "recv()" in "eval" if you don't want to die.

    If your $cv goes out of scope, then the request will be aborted.

  Callback
        $e->cluster_health->cb( sub {
            if ($@) {
                log "Error $@";
            }
            else {
                my $result = shift;
                log "$result"
            }
        })
        AE::cv->recv()

    If you set a callback on a CondVar, then the callback will be called
    once the CondVar is ready (which will only happen after you start the
    event loop).

    In the callback, $@ will contain any error, otherwise the result (if
    any) will be the first value in @_.

    Once you set a callback on a CondVar, it will not be aborted when it
    goes out of scope.

  Fire-and-Forget
        $e->delete_index(index=>'foo');

    If a request is called in "void" context, then it will be executed once
    the event loop is started. No errors will be thrown, even if the request
    does not complete succesfully.

    It will not be aborted with a change in scope, because there is no
    scope. If you exit the application without running an event loop, then
    any pending requests will not be run.

BLOCKING METHODS
    "scrolled_search()" in ElasticSearch and "reindex()" in ElasticSearch
    will be executed synchronously.

SEE ALSO
    *   ElasticSearch

    *   ElasticSearch::Transport

    *   ElasticSearch::Transport::HTTPLite

    *   ElasticSearch::Transport::HTTPTiny

    *   ElasticSearch::Transport::Curl

    *   ElasticSearch::Transport::AEHTTP

    *   ElasticSearch::Transport::AECurl

    *   ElasticSearch::Transport::Thrift

    1;

AUTHOR
    Clinton Gormley <drtech@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Clinton Gormley.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for ElasticSearch-Transport-AEHTTP

0.06    2012-11-14
        Archive was not being built correctly - copied ES test into this repo.

0.05    2012-11-07
        Added a max_content_length attribute to ElasticSearch::Transport
        to avoid posting too large a body to elasticsearch. The value
        is auto-detected in versions 0.19.12 onwards

0.04    2012-03-02
        Added skip_request to make short circuits work in bulk/mget/msearch

0.03    2012-02-17
        Use ElasticSearch::Transport::code_to_error to centralise HTTP error
        code mapping.
        Now requires ElasticSearch v 0.48

0.02    2011-08-27
        Added support for deflate()
        Fixed a bug that was causing ElasticSearch::Transport::AECurl to not
        release all of its requests on DESTROY

0.01    2011-08-26

        First release


t/request_tests/mget.pl
#!perl

use Test::More;
use strict;
use warnings;
our $es;
my $r;

ok $r= $es->mget(
    index => 'es_test_1',
    type  => 'type_1',
    ids   => [ 1, 2, 3, 4, 5 ]
    ),
    'mget';
is scalar @$r, 5, ' - 5 results';
is $r->[0]{_id},    1, ' - first id ok';
is $r->[3]{exists}, 0, "id 3 doesn't exist";

ok $r= $es->mget(
    index          => 'es_test_1',
    type           => 'type_1',
    ids            => [ 1, 2, 3, 4, 5 ],
    filter_missing => 1
    ),
    ' - filter missing';
is scalar @$r, 2, ' - missing filtered';

ok $r= $es->mget(
    docs => [
        { _index => 'es_test_1', _type => 'type_1', _id => 1 },
        { _index => 'es_test_1', _type => 'type_1', _id => 5 }
    ]
    ),
    ' - docs';

ok $r= $es->mget(
    fields => [ 'num', 'date' ],
    docs   => [
        { _index => 'es_test_1', _type => 'type_1', _id => 1 },
        {   _index => 'es_test_1',
            _type  => 'type_1',
            _id    => 5,
            fields => ['text']
        }
    ]
    ),
    ' - fields';

ok keys %{ $r->[0]{fields} } == 2
    && $r->[0]{fields}{num}
    && $r->[0]{fields}{date}, ' - default';

ok keys %{ $r->[1]{fields} } == 1 && $r->[1]{fields}{text}, ' - specific';

is_deeply $r = $es->mget( docs => [] ), [], ' - no docs';
is $r = $es->mget( docs => [], as_json => 1 ), "[]\n", ' - no docs json';

throws_ok { $es->mget( type => 'foo' ) } qr/Cannot specify a type for mget/,
    ' - type without index';
throws_ok { $es->mget( ids => [] ) } qr/Use of the ids param with mget/,
    ' - ids no index';
throws_ok { $es->mget( index => 'es_type_1', ids => [], docs => [] ) }
qr/Cannot specify both ids and docs/, ' - ids and docs';

1


t/request_tests/count.pl
#!perl

use Test::More;
use strict;
use warnings;
our $es;
my $r;

is $es->count( query => { term => { text => 'foo' } } )->{count}, 17,
    "Count: query";

is $es->count( term => { text => 'foo' } )->{count}, 17, "Count: term";

is $es->count( terms => { text => [ 'foo', 'bar' ] } )->{count}, 25,
    "Count: terms";
is $es->count( in => { text => [ 'foo', 'bar' ] } )->{count}, 25, "Count: in";

is $es->count( range => { num => { gte => 10, lte => 20 } } )
    ->{count}, 11,
    'Count: range';

is $es->count( range => { num => { gt => 10, lt => 20 } } )
    ->{count}, 9,
    'Count: range, gt/lt';

is $es->count( prefix => { text => 'ba' } )->{count}, 24, 'Count: prefix';

is $es->count( wildcard => { text => 'ba?' } )->{count}, 24,
    'Count: wildcard';

is $es->count( match_all => {} )->{count}, 29, 'Count: match_all';

is $es->count(
    query_string => {
        query         => 'foo AND bar AND -baz',
        default_field => 'text'
    }
    )->{count}, 4,
    'Count: query_string';

is $es->count(
    bool => {
        must =>
            [ { term => { text => 'foo' } }, { term => { text => 'bar' } } ]
    }
    )->{count}, 8,
    'Count: bool';

is $es->count(
    dis_max => {
        queries =>
            [ { term => { text => 'foo' } }, { term => { text => 'bar' } } ]
    }
    )->{count}, 25,
    'Count: dis_max';

is $es->count(
    constant_score => { filter => { terms => { text => [ 'foo', 'bar' ] } } }
    )->{count}, 25,
    'Count: constant_score';

is $es->count(
    filtered => {
        query  => { term => { text => 'foo' } },
        filter => { term => { text => 'bar' } }
    }
    )->{count}, 8,
    'Count: filtered';

is $es->count( field => { text => 'foo' } )->{count}, 17, 'Count: field';

is $es->count(
    fuzzy => {
        text => {
            value          => 'bart',
            prefix_length  => 1,
            min_similarity => 0.4
        }
    }
)->{count}, 16, 'Count: fuzzy';

is $es->count( flt => { fields => ['text'], like_text => 'bat' } )
    ->{count}, 24,
    'Count: fuzzy_like_this';

is $es->count( flt_field => { text => { like_text => 'fooo' } } )
    ->{count}, 17,
    'Count: fuzzy_like_this_field';

is $es->count(
    mlt => {
        like_text     => 'foo bar baz',
        min_term_freq => 1,
        min_doc_freq  => 1
    }
    )->{count}, 29,
    'Count: more_like_this';

is $es->count(
    mlt_field => {
        text => {
            like_text     => 'foo bar baz',
            min_term_freq => 1,
            min_doc_freq  => 1
        }
    }
    )->{count}, 29,
    'Count: more_like_this';

is $es->count(
    span_first => {
        match => {
            span_near => {
                clauses => [
                    { span_term => { text => 'baz' } },
                    { span_term => { text => 'bar' } }
                ],
                slop     => 0,
                in_order => 0
            }
        },
        end => 2
    }
)->{count}, 4, 'Count: span queries';

TODO: {
    local $TODO = "field_masking_span queries not recognised by server";
    is $es->count(
        field_masking_span => {
            field => 'num',
            query => { span_term => { text => 'foo' } }
        }
    )->{count}, 1111, 'Count: field_masking_span';
}

1


t/request_tests/rivers.pl
#!perl

use Test::More;
use strict;
use warnings;
our $es;
my $r;

no warnings 'redefine';

### RIVER ###
ok $es->create_river( type => 'dummy', river => 'foo' )->{ok}, 'Create river';
wait_for_es(2);
is $es->get_river( river => 'foo' )->{_type}, 'foo', 'Get river';
is $es->river_status( river => 'foo' )->{_id}, '_status', 'River status';
ok $es->delete_river( river => 'foo' )->{ok}, 'Delete river';

missing($_) for qw(get_river river_status);

$es->delete_index( index => '_river' );

#===================================
sub missing {
#===================================
    my $action = shift;
    throws_ok { $es->$action( river => 'foobar' ) }
    qr/ElasticSearch::Error::Missing/, " - $action missing";
    ok !$es->$action( river => 'foobar', ignore_missing => 1 ),
        " - $action ignore_missing";
}

1


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "AnyEvent::HTTP backend for ElasticSearch",
  "dist_author" => [
    "Clinton Gormley <drtech\@cpan.org>"
  ],
  "dist_name" => "ElasticSearch-Transport-AEHTTP",
  "dist_version" => "0.02",
  "license" => "perl",
  "module_name" => "ElasticSearch::Transport::AEHTTP",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "AnyEvent" => 0,
    "AnyEvent::HTTP" => 0,
    "ElasticSearch" => "0.44",
    "ElasticSearch::Transport" => 0,
    "ElasticSearch::Util" => 0,
    "Encode" => 0,
    "Scalar::Util" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "AnyEvent::HTTP backend for ElasticSearch",
  "AUTHOR" => "Clinton Gormley <drtech\@cpan.org>",
  "BUILD_REQUIRES" => {
    "ElasticSearch::SearchBuilder" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.96"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "ElasticSearch-Transport-AEHTTP",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "ElasticSearch::Transport::AEHTTP",
  "PREREQ_PM" => {
    "AnyEvent" => 0,
    "AnyEvent::HTTP" => 0,
    "ElasticSearch" => "0.60",
    "ElasticSearch::Transport" => 0,
    "ElasticSearch::Util" => 0,
    "Encode" => 0,
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.06",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/ElasticSearch/Transport/AEHTTP.pm
package ElasticSearch::Transport::AEHTTP;

use strict;
use warnings;

use ElasticSearch 0.60 ();
use parent 'ElasticSearch::Transport';
use AnyEvent::HTTP qw(http_request);
use Encode qw(decode_utf8 encode_utf8);
use ElasticSearch::Util qw(build_error);
use Scalar::Util qw(weaken isweak);
use List::Util qw(shuffle min);

our $VERSION = '0.06';

#===================================
sub protocol     {'http'}
sub default_port {9200}
#===================================

#===================================
sub request         { shift->_cv_wrap( '_request',         @_[ 0 .. 1 ] ) }
sub send_request    { shift->_cv_wrap( '_send_request',    @_ ) }
sub current_server  { shift->_cv_wrap( '_current_server',  @_ ) }
sub next_server     { shift->_cv_wrap( '_next_server',     @_ ) }
sub refresh_servers { shift->_cv_wrap( '_refresh_servers', @_ ) }
sub skip_request    { shift->_cv_wrap( '_skip_request',    @_ ) }
#===================================

#===================================
sub _request {
#===================================
    my $self   = shift;
    my $params = shift;
    my $s_srvr = shift;

    my $weak_cb = my $cb = shift;
    weaken $weak_cb;

    my $args = $self->_tidy_params($params);
    $self->reset_failed_servers();

    my $weak_req;
    my $request_cb = $weak_req = sub {
        my $srvr = shift
            or return $weak_cb->( undef, @_ );

        $self->log_request( $srvr, $args ) unless $s_srvr;

        $self->_send_request(
            $srvr, $args,
            sub {
                my $json = shift || '{"ok": true}';
                if ( my $error = shift ) {
                    if ( !$s_srvr && $self->should_retry( $srvr, $error ) ) {
                        my @guard;
                        my $next_cb = sub { $weak_req->(@_); @guard = () };
                        return @guard = $self->_next_server($next_cb);
                    }

                    $error = $self->_handle_error( $srvr, $params, $error );
                    return $weak_cb->( undef, $error );
                }
                local $@;
                return $weak_cb->(
                    eval { $self->_response( $json, $params, $s_srvr ) }
                        || undef,
                    $@
                );
            }
        );
    };

    weaken $weak_req;
    return ( $cb, $request_cb, $s_srvr
        ? $request_cb->($s_srvr)
        : $self->_next_server($request_cb) );

}

#===================================
sub _send_request {
#===================================
    my $self   = shift;
    my $server = shift;
    my $params = shift;
    my $cb     = shift;

    my $method = $params->{method};
    my $uri = $self->http_uri( $server, $params->{cmd}, $params->{qs} );

    my $data = $params->{data};
    if ( defined $data ) {
        $data = encode_utf8($data);
        eval { $self->check_content_length( \$data ); 1 }
            or $cb->( undef, $@ );
    }

    my $request_cb = sub {
        my ( $content, $hdr ) = @_;

        $content = '' unless defined $content;
        $content = $self->inflate($content)
            if ( $hdr->{'content-encoding'} || '' ) eq 'deflate';

        $content = decode_utf8($content);
        my $code = $hdr->{Status};

        if ( $code =~ /^2/ ) {
            return $cb->($content);
        }
        my $msg = $hdr->{Reason};

        my $type = $self->code_to_error($code)
            || (
              $msg eq 'Connection timed out' ? 'Timeout'
            : $msg =~ /Broken pipe|Connection (reset by peer|refused)/
            ? 'Connection'
            : 'Request'
            );

        my $error_params = {
            server      => $server,
            status_code => $code,
            status_msg  => $msg,
        };

        $error_params->{content} = $content
            unless $type eq 'Timeout';

        my $error = build_error( $self, $type, $msg, $error_params );
        $cb->( undef, $error );
    };

    my $headers;
    $headers = { 'Accept-Encoding' => 'deflate' }
        if $self->deflate;

    return http_request(
        $method    => $uri,
        body       => $data,
        timeout    => $self->timeout,
        persistent => 0,
        headers    => $headers,
        $request_cb
    );
}

#===================================
sub _skip_request {
#===================================
    my $self = shift;
    my $cb   = pop;
    $cb->( $self->SUPER::skip_request(@_) );
}

#===================================
sub _refresh_servers {
#===================================
    my $self = shift;
    my $cb   = shift;

    my $queue = $self->{_queue} ||= [];
    push @$queue, $cb;
    weaken $queue->[-1];

    my $requests = $self->{_requests} ||= [];
    weaken $self->{_requests} unless isweak $self->{_requests};
    @$requests = () unless grep $_, @$requests;
    return ( @$requests, $cb ) if ( grep $_, @$queue ) > 1;

    $self->{_refresh_in} = 0;
    delete $self->{_current_server};

    my %servers = map { $_ => 1 }
        ( @{ $self->servers }, @{ $self->default_servers } );

    my @all_servers = keys %servers;
    my $protocol    = $self->protocol;

    my $count         = @all_servers;
    my $process_queue = sub {
        my $ok = shift;
        my $error;
        if ( !$ok ) {
            return if --$count;
            $error = build_error(
                $self, 'NoServers',
                "Could not retrieve a list of active servers:\n$_[0]",
                { servers => \@all_servers }
            );
        }
        @$requests = ();
        while (@$queue) {
            my $cb = shift @$queue or next;
            $cb->( $ok, $error );
        }
    };

    my $request_cb = sub {
        my $nodes = shift() || return $process_queue->( undef, @_ );
        my @servers = grep {$_}
            map {m{/([^]]+)}}
            map {
                   $_->{ $protocol . '_address' }
                || $_->{ $protocol . 'Address' }
                || ''
            } values %{ $nodes->{nodes} };

        return $process_queue->(
            undef,
            build_error(
                $self, 'Internal',
                "ElasticSearch returned no live servers"
            )
        ) unless @servers;

        if ( $protocol eq 'http' ) {
            my $content_length = min( $self->max_content_length,
                grep {$_} map { $_->{http}{max_content_length_in_bytes} }
                    values %{ $nodes->{nodes} } );
            $self->max_content_length($content_length);
        }
        @servers = shuffle @servers;
        $self->servers( \@servers );
        $self->{_refresh_in} = $self->max_requests - 1;
        %servers = ();
        return $process_queue->( \@servers );
    };

    foreach my $server (@all_servers) {
        next unless $server;
        push @$requests,
            $self->_request(
            { cmd => '/_cluster/nodes', qs => { http => 1 } },
            $server, $request_cb );
    }

    my @requests = @$requests;
    weaken $requests->[$_] for 0 .. $#requests;

    return ( @$requests, $cb );
}

#===================================
sub _current_server {
#===================================
    my $self = shift;
    my $cb   = shift;
    return $cb->( $self->{_current_server}{$$} )
        if $self->{_current_server}{$$};

    return $self->_next_server($cb);
}

#===================================
sub _next_server {
#===================================
    my $self = shift;
    my $cb   = shift;
    local $@;

    return $cb->( eval { $self->SUPER::next_server() } || undef, $@ )
        if $self->{_refresh_in} || $self->no_refresh;

    return $self->_refresh_servers(
        sub {
            return $cb->(@_) if $_[1];
            $self->{_refresh_in}++;
            $cb->( eval { $self->SUPER::next_server() } || undef, $@ );
            $self->{_refresh_in}--;
        }
    );
}

#===================================
sub cv { shift; ElasticSearch::Transport::AnyEvent::CondVar->new(@_) }
#===================================

#===================================
sub _cv_wrap {
#===================================
    my $self    = shift;
    my $method  = shift;
    my $cv_weak = my $cv = $self->cv;
    weaken $cv_weak;

    $cv->guard(
        $self->$method(
            @_,
            sub {
                my $content = shift;
                $_[0]
                    ? $cv_weak->croak( shift() )
                    : $cv_weak->send($content);
                $cv_weak->clear_guard() if $cv_weak;
            }
        )
    );

    $cv->guard($cv) unless defined wantarray;
    return $cv;
}

#===================================
sub _make_sync {
#===================================
    my $class = shift;
    my $old   = \&_cv_wrap;
    no warnings 'redefine';
    *_cv_wrap   = sub { $old->(@_)->recv };
    *_make_sync = sub { };
}

#===================================
#===================================
package ElasticSearch::Transport::AnyEvent::CondVar;
#===================================
#===================================

use AnyEvent();
our @ISA = qw(AnyEvent::CondVar);

my $created = my $destroyed = 0;
our %CV;

#===================================
sub new {
#===================================
    my $proto = shift;
    my $class = ref $proto || $proto;
    my $self  = AnyEvent->condvar;
    bless $self, $class;
    $self->{_guard} = [];

    if ($ElasticSearch::DEBUG) {
        $created++;
        $self->{_line}
            = join( ' ', $created, ( caller(3) )[3], ( caller(2) )[2] );
        print $self->{_line} . "\n";
        $CV{ $self->{_line} } = $self;
        Scalar::Util::weaken $CV{ $self->{_line} };
    }
    return $self;
}

#===================================
sub guard {
#===================================
    my $self = shift;
    push @{ $self->{_guard} }, @_;
}

#===================================
sub clear_guard { shift->{_guard} = []; }
sub clear_cv { delete shift->{_ae_cv} }
#===================================

#===================================
sub cb {
#===================================
    my $self = shift;
    return $self->SUPER::cb unless @_;

    my $cb = shift;
    my $weak_cv = my $cv = $self;
    Scalar::Util::weaken $weak_cv;
    $self->guard($cv);
    $self->SUPER::cb(
        sub {
            local $@ = $weak_cv->{_ae_croak};
            $cb->( @{ $weak_cv->{_ae_sent} } );
        }
    );
}

#===================================
sub recv {
#===================================
    $_[0]->_wait;
    die $_[0]{_ae_croak} if $_[0]{_ae_croak};
    wantarray ? @{ $_[0]{_ae_sent} } : $_[0]{_ae_sent}[0];
}

#===================================
sub DESTROY {
#===================================
    my $self = shift;
    if ( $ElasticSearch::DEBUG && $self->{_line} ) {
        print "Destroyed: " . $self->{_line} . "\n";
        delete $CV{ $self->{_line} };
        $destroyed++;
    }
    $self->clear_guard;
}

#===================================
sub stats {
#===================================
    print "Created: $created\nDestroyed: $destroyed\n";
}
1;

# ABSTRACT: AnyEvent::HTTP backend for ElasticSearch


__END__
=pod

=head1 NAME

ElasticSearch::Transport::AEHTTP - AnyEvent::HTTP backend for ElasticSearch

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use ElasticSearch;
    my $e = ElasticSearch->new(
        servers     => 'search.foo.com:9200',
        transport   => 'aehttp',
    );

    # blocking request
    $e->cluster_health->recv;

    # non-blocking request
    $e->cluster_health->cb( sub {
        if ($@) {
            log "An error occurred: $@";
        } else {
            my $response = shift;
            log $response;
        }
    });
    AE::cv->recv();

    # fire-and-forget vs scoped
    {
        $e->delete_index(index => 'foo');
        $e->delete_index(index => 'bar')->cb( sub { print "Done"});
        my $cv = $e->delete_index(index=>'baz');
    }
    AE::cv->recv;
    #   - foo and bar will be deleted
    #   - baz will not be deleted

=head1 DESCRIPTION

ElasticSearch::Transport::AEHTTP uses L<AnyEvent::HTTP> to talk to
L<ElasticSearch> asynchronously over HTTP.

=head1 USING AEHTTP

Any request to ElasticSearch returns an L<AnyEvent::CondVar>. You have
three options for how you use them:

=head2 Blocking

    $cv = $e->cluster_health;
    $result = $cv->recv;

When you call C<recv()> on a CondVar, your program will block until
that CondVar is ready to return a value.

If an error was thrown, then C<recv()> will C<die>.  You will need to wrap
C<recv()> in C<eval> if you don't want to die.

If your C<$cv> goes out of scope, then the request will be aborted.

=head2 Callback

    $e->cluster_health->cb( sub {
        if ($@) {
            log "Error $@";
        }
        else {
            my $result = shift;
            log "$result"
        }
    })
    AE::cv->recv()

If you set a callback on a CondVar, then the callback will be called once
the CondVar is ready (which will only happen after you start the event loop).

In the callback, C<$@> will contain any error, otherwise the result (if any)
will be the first value in C<@_>.

Once you set a callback on a CondVar, it will not be aborted when it goes
out of scope.

=head2 Fire-and-Forget

    $e->delete_index(index=>'foo');

If a request is called in C<void> context, then it will be executed once
the event loop is started. No errors will be thrown, even if the request
does not complete succesfully.

It will not be aborted with a change in scope, because there is no scope. If
you exit the application without running an event loop, then any pending
requests will not be run.

=head1 BLOCKING METHODS

L<ElasticSearch/"scrolled_search()"> and L<ElasticSearch/"reindex()"> will
be executed synchronously.

=head1 SEE ALSO

=over

=item * L<ElasticSearch>

=item * L<ElasticSearch::Transport>

=item * L<ElasticSearch::Transport::HTTPLite>

=item * L<ElasticSearch::Transport::HTTPTiny>

=item * L<ElasticSearch::Transport::Curl>

=item * L<ElasticSearch::Transport::AEHTTP>

=item * L<ElasticSearch::Transport::AECurl>

=item * L<ElasticSearch::Transport::Thrift>

=back

1;

=head1 AUTHOR

Clinton Gormley <drtech@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Clinton Gormley.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

